### ISSUE TITLE
canDecodeType \ canEncodeType could be supports()

### ISSUE THREAD

**Issue by marcoscaceres** [WICG/img-conversion#12]
I'm wondering, why not: 

``` js
partial interface ImageData {
  static CanDoOperation supports (EncoderOpts options);
};
dictionary EncoderOpts{
  Operation operation;
  DOMString type = "image/png";
};
enum CanDoOperation{
  "", "maybe", "probably"
};
enum Operation {
  "encode", "decode"
};
```

Then:

``` js
var ops = {operation: "decode", type: "image/jpeg"};
if (ImageData.supports(ops)){
    //do thing..
}
```

**Comment by marcoscaceres** [comment]
That would allow for checking more extensive things if we add more to the `EncoderOpts` dictionary.

**Comment by marcoscaceres** [comment]
updated the IDL above to include `EncoderOpts`.

**Comment by marcoscaceres** [comment]
Following HTML doesn't matter in this case, this is also a Worker API.

**Comment by marcoscaceres** [comment]
Basically... I want to be able to do things like: 

``` JS
async function () {
  let ops = {
      type: "image/jpg",
      operation: "encode",
  }
  if (!(ImageData.supports(ops)) {
     ops.type = "image/png";
  }
  let blob = await imageData.toBlob(ops);
  const img = document.body.appendChild(new Image());
  img.src = URL.createObjectURL(blob);
}
```

**Comment by AshleyScirra** [comment]
What operations do you envisage ImageData supporting other than encode and decode? I'm not sure why that would need to be an option like that.

I think it is a good idea to design the platform to be consistent, and if it is only necessary to check support for a MIME type, it ought to be consistent with the functionality and naming of HTMLMediaElement.canPlayType and MediaRecorder.canRecordType.

**Comment by marcoscaceres** [comment]
Having spoken to our graphics team, we absolutely don't want to end up having to parse MIME strings to derive options, like with .canPlayType().

**Comment by marcoscaceres** [comment]
To be clear: 

> What operations do you envisage ImageData supporting other than encode and decode?

None. Just those two. 

> I'm not sure why that would need to be an option like that.

Just to reduce API surface area. 

> I think it is a good idea to design the platform to be consistent, and if it is only necessary to check support for a MIME type, it ought to be consistent with the functionality and naming of HTMLMediaElement.canPlayType and MediaRecorder.canRecordType.

I don't disagree with this in principle, but it's ok to move away from things like the above where alternatives make sense. I.e., it's ok to break from convention, like with SW's `event.waitUntil()`.

**Comment by AshleyScirra** [comment]
If the graphics team don't want to have to parse MIME strings, what's the alternative? Should we have something like `supports("webp")` or `supports({type: "webp"})`? That looks like reinventing the wheel when MIME types already exist to describe these formats.

**Comment by marcoscaceres** [comment]
It's fine to have just "image/whatever"... we just don't want to get into a situation where we end up with things like:

```
'image/vnd.microsoft.icon;format=bmp,bpp=32'
```

Or similar.

**Comment by AshleyScirra** [comment]
Yeah, that would suck. Perhaps the spec could mandate that only MIME types of the form "image/XXX" are allowed (with no extra parameters)? If the support is mixed depending on parameters, the functions are allowed to return "maybe". The situation is similar to HTMLMediaElement's `canPlayType` which you can pass the MIME type for a WAV file which in theory could have an exotic sample rate/bit depth/etc, but in common usage will work. AFAIK no browsers bother to parse actual parameters like the sample rate or bit depth in that case, they just return "maybe" instead of "probably".
